crates/aptos-admin-service/src/server/mod.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::server::utils::reply_with_status;
+use aptos_config::config::{AuthenticationConfig, NodeConfig};
+use aptos_consensus::{
+    persistent_liveness_storage::StorageWriteProxy, quorum_store::quorum_store_db::QuorumStoreDB,
+};
+use aptos_infallible::RwLock;
+use aptos_logger::info;
+use aptos_storage_interface::DbReaderWriter;
+use hyper::{
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server, StatusCode,
+};
+use std::{
+    collections::HashMap,
+    convert::Infallible,
+    net::{SocketAddr, ToSocketAddrs},
+    sync::Arc,
+};
+use tokio::runtime::Runtime;
+
+mod consensus;
+#[cfg(target_os = "linux")]
+mod profiling;
+#[cfg(target_os = "linux")]
+mod thread_dump;
+mod utils;
+
+#[derive(Default)]
+pub struct Context {
+    authentication_configs: Vec<AuthenticationConfig>,
+
+    aptos_db: RwLock<Option<Arc<DbReaderWriter>>>,
+    consensus_db: RwLock<Option<Arc<StorageWriteProxy>>>,
+    quorum_store_db: RwLock<Option<Arc<QuorumStoreDB>>>,
+}
+
+impl Context {
+    fn set_aptos_db(&self, aptos_db: Arc<DbReaderWriter>) {
+        *self.aptos_db.write() = Some(aptos_db);
+    }
+
+    fn set_consensus_dbs(
+        &self,
+        consensus_db: Arc<StorageWriteProxy>,
+        quorum_store_db: Arc<QuorumStoreDB>,
+    ) {
+        *self.consensus_db.write() = Some(consensus_db);
+        *self.quorum_store_db.write() = Some(quorum_store_db);
+    }
+}
+
+pub struct AdminService {
+    runtime: Runtime,
+    context: Arc<Context>,
+}
+
+impl AdminService {
+    /// Starts the admin service that listens on the configured address and handles various endpoint
+    /// requests.
+    pub fn new(node_config: &NodeConfig) -> Self {
+        // Fetch the service port and address
+        let service_port = node_config.admin_service.port;
+        let service_address = node_config.admin_service.address.clone();
+
+        // Create the admin service socket address
+        let address: SocketAddr = (service_address.as_str(), service_port)
+            .to_socket_addrs()
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Failed to parse {}:{} as address",
+                    service_address, service_port
+                )
+            })
+            .next()
+            .unwrap();
+
+        // Create a runtime for the admin service
+        let runtime = aptos_runtimes::spawn_named_runtime("admin".into(), None);
+
+        let admin_service = Self {
+            runtime,
+            context: Arc::new(Context {
+                authentication_configs: node_config.admin_service.authentication_configs.clone(),
+                ..Default::default()
+            }),
+        };
+
+        // TODO(grao): Consider support enabling the service through an authenticated request.
+        let enabled = node_config.admin_service.enabled.unwrap_or(false);
+        admin_service.start(address, enabled);
```

Comment:
hmm, why enable = false we still create the service but just return not found?

Reply:
When your enter a url and your browser doesn't load anything, it's hard to tell if it's because of the service not enabled, or other errors. Explicitly returns a page will make it easier to understand what's happening.